server/src/main/java/org/elasticsearch/action/support/IndicesOptions.java

```diff
@@ -22,138 +24,425 @@
+                matchOpen = false;
+                matchClosed = false;
+                includeHidden = false;
+                return this;
+            }
+
+            /**
+             * Maximises the resolution of indices, we will match open, closed and hidden targets.
+             */
+            public Builder all() {
+                matchOpen = true;
+                matchClosed = true;
+                includeHidden = true;
+                return this;
+            }
+
+            /**
+             * Parses the list of wildcard states to expand as provided by the user.
+             * Logs a warning when the option 'none' is used along with other options because the position in the list
+             * changes the outcome.
+             */
+            public Builder expandStates(String[] expandStates) {
+                // Calling none() simulates a user providing an empty set of states
+                none();
+                for (String expandState : expandStates) {
+                    switch (expandState) {
+                        case "open" -> matchOpen(true);
+                        case "closed" -> matchClosed(true);
+                        case "hidden" -> includeHidden(true);
+                        case "all" -> all();
+                        case "none" -> {
+                            none();
+                            if (expandStates.length > 1) {
+                                DEPRECATION_LOGGER.warn(DeprecationCategory.API, "expand_wildcards", WILDCARD_NONE_DEPRECATION_MESSAGE);
+                            }
+                        }
+                    }
+                }
+                return this;
             }
+
+            public WildcardOptions build() {
+                return new WildcardOptions(matchOpen, matchClosed, includeHidden, resolveAliases, allowEmptyExpressions);
+            }
+        }
+
+        public static Builder newBuilder() {
+            return new Builder();
+        }
+
+        public static Builder newBuilder(WildcardOptions wildcardOptions) {
+            return new Builder(wildcardOptions);
         }
     }
 
-    public enum Option {
-        IGNORE_UNAVAILABLE,
-        IGNORE_ALIASES,
-        ALLOW_NO_INDICES,
-        FORBID_ALIASES_TO_MULTIPLE_INDICES,
-        FORBID_CLOSED_INDICES,
-        IGNORE_THROTTLED;
+    /**
+     * These options apply on all indices that have been selected by the other Options. It can either filter the response or
+     * define what type of indices or aliases are not allowed which will result in an error response.
+     * @param allowAliasToMultipleIndices, allow aliases to multiple indices, true by default.
+     * @param allowClosedIndices, allow closed indices, true by default.
+     * @param removeThrottled, filters out throttled (aka frozen indices), defaults to true.
+     */
+    public record GeneralOptions(boolean allowAliasToMultipleIndices, boolean allowClosedIndices, @Deprecated boolean removeThrottled)
```

Comment:
I think `ignoreThrottled` is clearer than "remove". If it were easier I'd prefer to invert it and have it be `includeThrottled`, but that is probably not worth the effort for BWC since this is already deprecated.